src/parser/paladin/retribution/modules/talents/DivinePurpose.js

```diff
@@ -1,43 +1,96 @@
+
 class DivinePurpose extends Analyzer {
   divinePurposeProcs = 0;
+  templarsVerdictConsumptions = 0;
+  divineStormConsumptions = 0;
+  justicarsVengeanceConsumptions = 0;
+  largestProcChain = 0;
+  currentProcChain = 0;
+  lastDivinePurposeConsumption = null;
 
   constructor(...args) {
     super(...args);
-    const hasDivinePurpose = this.selectedCombatant.hasTalent(SPELLS.DIVINE_PURPOSE_TALENT_RETRIBUTION.id);
-    this.active = hasDivinePurpose;
+    this.active = this.selectedCombatant.hasTalent(SPELLS.DIVINE_PURPOSE_TALENT_RETRIBUTION.id);
+  
+    // event listeners
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER).spell(HOLY_POWER_SPENDERS), this.onCast);
+    this.addEventListener(Events.applybuff.by(SELECTED_PLAYER).spell(SPELLS.DIVINE_PURPOSE_BUFF), this.onApplyDivinePurpose);
   }
 
-  on_byPlayer_applybuff(event) {
+  onCast(event){
     const spellId = event.ability.guid;
-    if (spellId === SPELLS.DIVINE_PURPOSE_BUFF.id) {
-      this.divinePurposeProcs++;
+    if (!this.selectedCombatant.hasBuff(SPELLS.DIVINE_PURPOSE_BUFF.id)) {
+      return;
+    }  
+    switch (spellId) {
+      case SPELLS.TEMPLARS_VERDICT.id:
+        this.templarsVerdictConsumptions += 1;
+        break;
+      case SPELLS.DIVINE_STORM.id:
+      // Empyrean Power grants a free Divine Storm and is prioritized over Divine Purpose if both are active
+        if (this.selectedCombatant.hasBuff(SPELLS.EMPYREAN_POWER_BUFF.id)) {
+          return;
+        }
+        this.divineStormConsumptions += 1;
+        break;
+      case SPELLS.JUSTICARS_VENGEANCE_TALENT.id:
+        this.justicarsVengeanceConsumptions += 1;
+        break;
+      default:
+        break;
     }
+    this.lastDivinePurposeConsumption = event.timestamp;
   }
 
-  on_byPlayer_refreshbuff(event) {
-    const spellId = event.ability.guid;
-    if (spellId === SPELLS.DIVINE_PURPOSE_BUFF.id) {
-      this.divinePurposeProcs++;
+  onApplyDivinePurpose(event){
+    this.divinePurposeProcs++;
+    const timeSinceLastDPconsumption = event.timestamp - this.lastDivinePurposeConsumption;
+    if (timeSinceLastDPconsumption < CHAIN_PROC_BUFFER) {
+      if (this.currentProcChain === 0) {
+        this.currentProcChain = 2;
+      }
+      else {
+        this.currentProcChain += 1;
+      }
+      if (this.currentProcChain > this.largestProcChain){
+        this.largestProcChain = this.currentProcChain;
+      }
+    }
+    else {
+      this.currentProcChain = 0;
     }
   }
 
   statistic() {
+    const hasJV = this.selectedCombatant.hasTalent(SPELLS.JUSTICARS_VENGEANCE_TALENT.id);
     return (
       <StatisticBox
         position={STATISTIC_ORDER.OPTIONAL(1)}
         icon={<SpellIcon id={SPELLS.DIVINE_PURPOSE_TALENT_RETRIBUTION.id} />}
         value={`${formatNumber(this.divinePurposeProcs)}`}
-        label="Divine Purpose Procs"
+        label="Divine Purpose procs"
```

Comment:
Please consider using the same statistic I used for the Holy Paladin DP stat, as I think it presents the data slightly more informatively.

Reply:
Yeah, i will take a look at trying that out, i don't like hiding a lot of the information in the tooltip either